[HOP][inductor] Add higher order associative scan operator

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122137
* __->__ #119430

Currently only supports single tensor scans, e.g. `cumsum`, `cumprod`, `logcumsumexp`


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

peterbell10 committed:
WIP: Add higher order associative scan operator

[ghstack-poisoned]

vadimkantorov commented:
It would also be nice to somehow support non-pointwise combine ops:

- like full-rank linear RNNs / some complex numbers/quaternion-valued/mat4x4 processing et cetera

For Mamba, it's sufficient to support only pointwise combine as it relies on diagonal transfer matrices, but it need not to be the case, and probably in future some more complex/expressive interactions might win back.

lezcano reviewed (commented):
It'd be nice if the user didn't need to provide an identity. It looks like the identity is just used to trace the op in proxy mode. Could we use the `input` instead there and remove the identity? We could have an early return if the dimension to scan over is <= 1, and pass the first two elements of the dimension to the function otherwise.

peterbell10 committed:
Update on "WIP: Add higher order associative scan operator"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Add higher order associative scan operator"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Add higher order associative scan operator"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Add higher order associative scan operator"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aakhundov commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -0,0 +1,173 @@
+
+@associative_scan_op.py_impl(DispatchKey.CompositeExplicitAutograd)
+def associative_scan_op_dense(input, dim, combine_fn):
+    raise NotImplementedError("associative_scan is not implemented for eager")
```
Wondering if this op fundamentally can't be implemented in eager or this is temporary? In principle, we could fallback to a serialized version in terms of eager ops?

peterbell10 replied:
Yes there's no reason we couldn't implement this in eager with pointwise operations and shuffling the tensor, it just wouldn't perform as well.

aakhundov commented:
@peterbell10 thanks for working on this! I've added some initial subgraph lowering and (inlined) subgraph codegen-ing infra in https://github.com/pytorch/pytorch/pull/119759. Do you think it would be possible to reuse this infra (or maybe parts of it) in this PR?

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Move fixup to right PR

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

lezcano commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -870,6 +872,123 @@ def _check_supported_callable(fn_var):
+        combine_graph = subtracer.graph
+        combine_graph.lint()
+
+        if subtracer.lifted_freevars:
+            unimplemented(
+                f"Combine fn had unexpected freevars: {subtracer.lifted_freevars}"
+            )
```
Can we graphbreak at all here? Probably better throw a hard error. Same below.

peterbell10 replied:
This function is decorated with `@raise_error_if_graph_break` so this does result in a hard error.

lezcano replied:
Why not simply raise an error directly?

peterbell10 replied:
This is consistent with how it's written in cond and while_loop. I suppose one reason to do it this way is all errors will give a similar error message no matter where in the graph capture process it errors.

lezcano reviewed (commented):
Looks very good. I didn't look at the registrations under `_higher_order_ops` tho.

peterbell10 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -870,6 +872,123 @@ def _check_supported_callable(fn_var):
+        with tx.fake_mode:
+            example_vals = [
+                torch.full(
+                    (), float("nan"), device=input_meta.device, dtype=input_meta.dtype
```
This is creating fake tensors so the value isn't actually used.

lezcano replied:
I guess you test integer inputs in the next PR with cummax so fair enough.

peterbell10 commented on torch/_inductor/subgraph_lowering.py:
```diff
@@ -0,0 +1,142 @@
+        return ops.placeholder(input_idx)
+
+    graph_inputs = [
+        ir.Pointwise.create(
```
I suppose I could make it an `InputBuffer` and override `ops.load` to access the placeholder value, but this works equally well and I don't need to define a layout. It also means I can detect if the code tries to realize the pointwise op, sincethe code would try to register a buffer for it.

peterbell10 commented on torch/_inductor/subgraph_lowering.py:
```diff
@@ -0,0 +1,142 @@
+
+    @property
+    def sizevars(self):
+        return self.root_graph.sizevars
```
Yup logging output confirms dynamic shapes are in the code.

peterbell10 commented on torch/_inductor/ops_handler.py:
```diff
@@ -709,3 +719,45 @@ def constant(self, value: Any, dtype: torch.dtype) -> "torch._inductor.ir.Consta
+    doesn't support stores which require load cache invalidation.
+    """
+
+    cse_cache: Dict[str, Union[T, Tuple[T, ...]]]
+    mock: MockHandler
```
This is just the syntax for typing attributes

https://mypy.readthedocs.io/en/stable/class_basics.html#instance-and-class-attributes

lezcano replied:
ugh. Can we do it like `self.x: list[int] = []` in the constructor to avoid the duplication?

peterbell10 committed:
Update

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -0,0 +1,173 @@
+aten = torch._ops.ops.aten
+
+
+def associative_scan(
+    input: torch.Tensor,
+    dim: int,
+    combine_fn: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+) -> torch.Tensor:
```
I get that this is meant to look like triton.associative_scan, but the rest of our control flow ops API follow JAX's API where the function comes first: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.associative_scan.html

peterbell10 replied:
Changed

zou3519 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -0,0 +1,173 @@
+        Read more about feature classification at:
+        https://pytorch.org/blog/pytorch-feature-classification-changes/#prototype
+
+    This operator requires runtime code generation and so requires support for
+    ``torch.compile``. Further, only CUDA device codegen is supported at the moment.
```
Our other control flow ops execute some python version of the op when torch.compile isn't available. Is there a reason to diverge?

peterbell10 replied:
An inefficient eager implementation is possible, but hasn't been written yet.

zou3519 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -0,0 +1,173 @@
+    pre_dispatch = getattr(proxy_mode, "pre_dispatch", False)
+
+    with disable_proxy_modes_tracing():
+        sample_inputs = (
+            torch.full((), False, dtype=input.dtype, device=input.device),
+            torch.full((), False, dtype=input.dtype, device=input.device),
+        )
```
There are two kinds of associative scan, right? One where the function takes in two scalars, and another where the function takes in two tensors. It sounds like this is the former?

peterbell10 replied:
Yes, this only supports scalar scans.

zou3519 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -0,0 +1,173 @@
+
+
+@associative_scan_op.py_functionalize_impl
+def associative_scan_functionalize(ctx, input, dim, combine_fn):
+    unwrapped_input = ctx.unwrap_tensors(input)
+    with ctx.redispatch_to_next() as m:
+        ret = associative_scan_op(unwrapped_input, dim, combine_fn)
+    return ctx.wrap_tensors(ret)
```
Are we saying that the combine_fn must not mutate any tensors?

peterbell10 replied:
Yes I don't think it makes sense to support mutations inside the combine_fn.

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -870,6 +872,120 @@ def _check_supported_callable(fn_var):
+                for _ in range(2)
+            ]
+
+        # Trace the subgraph
+        # NOTE: We don't use speculate_subgraph as it expects subgraph arguments to be
+        # variables in the current graph.
+        with tx.output.subtracer(self.value, prior_tracer=None) as subtracer:
```
I would recommend using speculate_subgraph (or figuring out how to get speculate_subgraph to work). speculate_subgraph has some additional things (and we'll continue to build on top of it):
- side effect restoration: https://github.com/pytorch/pytorch/blob/8c7d8f0ff20b71a2dc1fbe976ee457267ab385a9/torch/_dynamo/variables/higher_order_ops.py#L394
- pytree handling

Chillee replied:
This is how we're doing it for score_mod: https://github.com/pytorch/pytorch/pull/121845/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR1306

Chillee replied:
I was actually unaware it would produce extra junk 🤔 Is that right @drisspg ?

drisspg replied:
I have only been looking at at the joint graph in aot which seemed to be correct to me - with no extra nodes, this may be the case though for what dynamo produces

zou3519 replied:
It produces extra junk, but:
1) it gets cleaned up later on in the PT2 stack
2) we can also just go back in and remove it manually.

In general we want to centralize on calling speculate subgraph (or something else!) so that future infra improvements to the subgraph capture mechanism will benefit all users.

peterbell10 replied:
Okay, I've changed it to use `speculate_subgraph`.

peterbell10 committed:
Use speculate_subgraph and reorder args on "[HOP][inductor] Add higher order associative scan operator"


Currently only supports single tensor scans, e.g. `cumsum`, `cumprod`, `logcumsumexp`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[HOP][inductor] Add higher order associative scan operator"


Currently only supports single tensor scans, e.g. `cumsum`, `cumprod`, `logcumsumexp`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[HOP][inductor] Add higher order associative scan operator"


Currently only supports single tensor scans, e.g. `cumsum`, `cumprod`, `logcumsumexp`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -889,6 +891,106 @@ def _check_supported_callable(fn_var):
+
+        if combine_result_meta.shape != ():
+            unimplemented(
+                f"Expected cond_fn to return a tensor with shape () but got {combine_result_meta.shape}"
```
nit: cond_fn -> combine_fn

peterbell10 committed:
Update

[ghstack-poisoned]

Chillee reviewed (commented):
Overall, very cool!

One thing I want to discuss is what we're going to do about the "scalar" vs. "tensor"-level scan. My previous thinking was that we could do something like

```
associative_scan(..., mode="scalar")
```
and 
```
associative_scan(..., mode="tensor")
```
and this would correspond to the two different semantics for associative scan.

Wdyt?

Chillee commented:
@lezcano To clarify, by "tensor" vs. "scalar" semantics, I mean "full size tensor" vs. "scalar tensor". Of course we always want to use tensor operators on them.

>  if it gets to support things like scanning a batch of matrices and compute their cumulative product, like in an RNN

The thing is, we can also support this today without Triton support. Simply implement your blelloch scan at the PyTorch operator level. I think that's an implementation strategy that also makes a lot of sense, which is why I think we should have different modes for each one.

peterbell10 commented:
When lowering the subgraph to triton IR I currently check that every operator called is a pointwise operation. We could potentially detect the non-pointwise operator and automatically use the blelloch scan instead of raising an error. That would mean we could extend the current interface to support non-scalar scans without adding a new argument.

Chillee commented:
@peterbell10 I think it's tricky since the semantics can be totally different. 

e.g. today, when you're tracing the subgraph you're assuming that they're scalar tensors. However, the shapes that are being passed into the Tensor API should be "size of input tensor - scanned over dim"

peterbell10 commented:
I don't think the shape matters at all for the scalar `associative_scan` so it should be fine to always trace with the input tensor shape. Or is there something I'm missing?

Chillee commented:
I mean the function that you're tracing: https://github.com/pytorch/pytorch/pull/119430/files#diff-d6c3a57794a9a8a79eacedc57d68ff631fe80776f8701398eb3ef349973b033bR928

Specifically, the shapes here are `()`, or scalar tensors.

For example, let's say the user had some code like
```
def combine(a, b):
    return torch.mm(a, b)
```

This would error, as `a` and `b` both have `dim=0`, when `torch.mm` expects tensors with `dim=2`.

peterbell10 commented:
Yes, I'm saying the scalar associative scan should work the same no matter what the tensor shape is so I could change the subgraph tracer to use tensors with the same shape as the inputs and it would still work for simple scalar scan ops.  The `matmul` example would still error out for now as it's not a pointwise operator.

Chillee commented:
@peterbell10 I do think you still want a mode to be explicit.

I agree that we can automate it eventually, so that we always trace it with the "full" shape and then automatically determine if we can lower the scan into the kernel.

I'd note that it's more complicated than just "are there pointwise ops". For example, what should we lower to in the following situations?

```
c: [B]
def combine(a: [A, B], b: [A, B]):
    return (a + b) + c

c: [A]
def combine(a: [A, B], b: [A, B]):
    return (a + b) + c.unsqueeze(-1)

c: [A, B]
def combine(a: [A, B], b: [A, B]):
    return (a + b) + c
```

peterbell10 commented:
> I'd note that it's more complicated than just "are there pointwise ops". For example, what should we lower to in the following situations?

The subgraph tracing doesn't allow lifted free variables, so none of these are allowed currently and will raise this error:

```
Unsupported: Combine fn had unexpected freevars: {Proxy(g_import_main_c): Proxy(g_import_main_c)}
```

However, if we wanted to support this use case then it would have to be via the tensor version since we can't have a tensor inside of a triton level `associative_scan`.

I would note though that even for the tensor version the `combine_fn` has to be elementwise along the scan dimension (true for jax's api as well) which would mean that `(a + b) + c` must have `c` be a constant with respect to a single scan and is equivalent to a scalar scan.

Really the scalar scan is just an optimization of the tensor scan, rather than being a semantically different thing.

Chillee commented:
> However, if we wanted to support this use case then it would have to be via the tensor version since we can't have a tensor inside of a triton level associative_scan.

I mean, imo, this is another case where we would want the tensor version, right? If you're doing the scalar codegen you can't close over tensors but if you're doing the tensor codegen you can.

Overall, is your objection that you don't think we need an explicit mode about whether it's "scalar-level" vs. "tensor-level", and you think it's better to just automatically determine it?

IMO, it's better to start off explicit and then automatically determine it later if we decide that's feasible. It's easier to make that transition than the other way around.

lezcano commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -889,6 +891,106 @@ def _check_supported_callable(fn_var):
+                + f"got {combine_result_meta.dtype}"
+            )
+
+        if combine_result_meta.shape != ():
```
It'd be great to have a very weak check with `atol=0.1` or so, where we check that the function is associative using the first 3 elements of the tensor (if it has at least 3 in `dim`). This would be *super* useful to catch bugs when implementing non-trivial scans.

lezcano replied:
I was thinking where we would put this and there are a few possibilities:
1. Do it on a best effort basis: Create a 0-D tensor, potentially on CPU, and to assert_close. If there is any error when executing the ops, just skip the test.
2. Hide it behind a debug flag and advertise the flag in the docs. Then when this flag is on, we would realize the inputs to the scan and add an assert in the wrapper code. Otherwise, we could write a device_assert inside the code (provided slices already work in triton).

Probably option 1 (which we could also hide behind a flag, if you think that we may hit false positives?) may be already fine and would catch enough errors early on, being able to throw a nice error.

Chillee commented:
> For example, the eager implementation would always just be the tensor implementation, so the mode="scalar" version would still need to be able to call the combine function with full sized tensors.

I mean, you could just vmap it when you run in eager mode :)

> Yeah what I'm saying is that it's trivial to detect this case when lowering the subgraph, so not really an issue.

I think my concern still applies - the mental model for users is somewhat nontrivial.

Overall, though, if you feel strongly that this should be transparent to users, then I'm not against it.

Chillee commented on torch/_inductor/ops_handler.py:
```diff
@@ -712,3 +722,42 @@ def constant(self, value: Any, dtype: torch.dtype) -> "torch._inductor.ir.Consta
     return h
+
+
+class SimpleCSEHandler(WrapperHandler[T]):
```
Why do you need this instead of the normal CSE handler?

peterbell10 replied:
So this isn't strictly necessary. We would just toss the unfiltered `inner_fn`s into a closure and leave it for codegen to clean up.

I was thinking at some point we might want a real fx graph so we could do pattern matching or value range analysis or what have you, but not at the moment.

Chillee commented on torch/_inductor/subgraph_lowering.py:
```diff
@@ -0,0 +1,142 @@
+        )
+
+
+def lower_pointwise_subgraph(subgraph: ir.Subgraph, inputs: List[InputDescriptor]):
```
Not against it, but why is this needed? For templated attention we just iterated and manually called lowerings - is the main reason to check various kinds of safety issues?

peterbell10 replied:
Yes it's mainly for safety, I don't want the lowerings being called to interact with the top level graph in any way. So I isolate it by overriding `V.graph`.

For templated attention if it's not a pure pointwise operation then this function won't work, but the  `PointwiseSubgraphLowering` class could easily be adapted into `SubgraphLowering` that isn't as strict about what functions are allowed.

Chillee replied:
I mean, it's an indirect load, which is kind of a pointwise function?

peterbell10 replied:
Yeah there's no conceptual problem, it's just that I can't lower an `ops.load` inside a helper function in triton so want to ban such cases in this functions.

Chillee reviewed (approved):
Overall LGTM!

Chillee commented on test/inductor/test_torchinductor.py:
```diff
@@ -1359,6 +1359,31 @@ def fn(a, b):
         self.common(fn, (a, b), atol=1e-5, rtol=1e-5, check_lowp=False)
 
+    @skipCUDAIf(TEST_WITH_ROCM, "associative_scan is not supported on ROCm")
+    def test_custom_scan_op(self):
```
Also add a test where associative_scan is called within a larger function that's compiled.

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge

peterbell10 committed:
Update on "[HOP][inductor] Add higher order associative scan operator"


Currently only supports single tensor scans, e.g. `cumsum`, `cumprod`, `logcumsumexp`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

kit1980 commented:
Looks like this regressed
> sebotnet33ts_256                    FAIL:     accuracy=fail_accuracy, expected=pass

https://github.com/pytorch/pytorch/actions/runs/8802392831/job/24160531288

kit1980 commented:
Going to revert soon.

Chillee commented:
@kit1980 tbh, I would guess that there's a high chance of that being spurious/unrelated. This PR shouldn't touch any codegen for existing models AFAIK.

kit1980 commented:
OK, it started to pass after 3 failures in a row.
No need to revert.
